opm/autodiff/GraphColoring.hpp

```diff
@@ -0,0 +1,244 @@
+    std::vector<int> degrees(noVertices, 0);
+    int maxDegree = 0;
+    std::ptrdiff_t firstDegreeChange = 0;
+
+    // populate deque
+    for( auto vertex = graph.begin(), endVertex = graph.end();
+         vertex != endVertex; ++vertex)
+    {
+        auto currentVertex = *vertex;
+        auto& degree = degrees[currentVertex];
+
+        for(auto edge = graph.beginEdges(currentVertex),
+                endEdge = graph.endEdges(currentVertex);
+            edge != endEdge; ++edge)
+        {
+            ++degree;
+        }
+
+
+        if( degree >= maxDegree )
+        {
+            orderedVertices.emplace_front(currentVertex);
+            ++firstDegreeChange;
+            if(degree > maxDegree)
+            {
+                firstDegreeChange = 1;
+                maxDegree = degree;
+            }
+        }
+        else
+        {
+            orderedVertices.emplace_back(currentVertex);
+        }
+    }
+
+    // order deque by descending degree
+    std::stable_sort(orderedVertices.begin() + firstDegreeChange,
+                     orderedVertices.end(),
+              [&degrees](const Vertex& v1, const Vertex& v2)
+              {
+                  return degrees[v1] > degrees[v2];
+              });
+
+    // Overwrite degree with color
+    auto& colors = degrees;
+    std::fill(colors.begin(), colors.end(), -1);
+
+    int color = 0;
+    std::vector<std::size_t> verticesPerColor;
+    verticesPerColor.reserve(10);
+
+    while(!orderedVertices.empty())
+    {
+        verticesPerColor
+            .push_back(Detail::colorGraphWelshPowell(graph, orderedVertices, colors,
+                                                     color++, noVertices));
+    }
+    return std::make_tuple(colors, color, verticesPerColor);
+}
+
+/// \! Reorder colored graph preserving order of vertices with the same color.
+template<class Graph>
+std::vector<std::size_t>
+reorderVerticesPreserving(const std::vector<int>& colors, int noColors,
+                          const std::vector<std::size_t> verticesPerColor,
+                          const Graph& graph)
+{
+    std::vector<std::size_t> colorIndex(noColors, 0);
+    std::vector<std::size_t> indices(graph.maxVertex() + 1);
+    std::partial_sum(verticesPerColor.begin(),
+                     verticesPerColor.begin()+verticesPerColor.size() - 1,
+                     colorIndex.begin() + 1);
+
+    for(const auto& vertex: graph)
+    {
+        indices[vertex] = colorIndex[colors[vertex]]++;
+    }
+    return indices;
+}
+
+/// \! Reorder Vetrices in spheres
+template<class Graph>
+std::vector<std::size_t>
+reorderVerticesSpheres(const std::vector<int>& colors, int noColors,
+                       const std::vector<std::size_t> verticesPerColor,
+                       const Graph& graph,
+                       typename Graph::VertexDescriptor root)
+{
+    std::vector<std::size_t> colorIndex(noColors, 0);
+    const auto notVisitedTag = std::numeric_limits<std::size_t>::max();
```

Comment:
Pedantic: You need to include `<limits>` in order to use `numeric_limits<T>::max()`.